imcache_test.go

```diff
@@ -1352,3 +1356,231 @@ func TestCache_MaxEntriesLimit_LessOrEqual0(t *testing.T) {
+				c.Set("foo1", "bar", WithExpirationDate(time.Now().Add(-1*time.Second)))
+				c.ReplaceKey("foo1", "foo2", WithNoExpiration())
+			},
+		},
+		{
+			name: "ReplaceKey evict expired entry under new key",
+			execute: func(c imcache[string, string]) {
+				c.Set("bar", "foo", WithNoExpiration())
+				c.Set("foo", "bar", WithExpirationDate(time.Now().Add(-1*time.Second)))
+				c.ReplaceKey("bar", "foo", WithNoExpiration())
+			},
+		},
+		{
+			name: "ReplaceKey evict replaced entry under new key",
+			execute: func(c imcache[string, string]) {
+				c.Set("bar", "foo", WithNoExpiration())
+				c.Set("foo", "bar", WithNoExpiration())
+				c.ReplaceKey("bar", "foo", WithNoExpiration())
+			},
+		},
+		{
+			name: "ReplaceKey evict replaced key entry",
+			execute: func(c imcache[string, string]) {
+				c.Set("bar", "foo", WithNoExpiration())
+				c.ReplaceKey("bar", "foo", WithNoExpiration())
+			},
+		},
+		{
+			name: "Remove evict expired entry",
+			execute: func(c imcache[string, string]) {
+				c.Set("bar", "foo", WithExpirationDate(time.Now().Add(-1*time.Second)))
+				c.Remove("bar")
+			},
+		},
+		{
+			name: "Remove evict removed entry",
+			execute: func(c imcache[string, string]) {
+				c.Set("bar", "foo", WithNoExpiration())
+				c.Remove("bar")
+			},
+		},
+		{
+			name: "RemoveAll evict removed and expired entries",
+			execute: func(c imcache[string, string]) {
+				c.Set("bar", "foo", WithNoExpiration())
+				c.Set("foo", "bar", WithExpirationDate(time.Now().Add(-1*time.Second)))
+				c.RemoveAll()
+			},
+		},
+		{
+			name: "RemoveExpired evict expired entry",
+			execute: func(c imcache[string, string]) {
+				c.Set("bar", "foo", WithNoExpiration())
+				c.Set("foo", "bar", WithExpirationDate(time.Now().Add(-1*time.Second)))
+				c.RemoveExpired()
+			},
+		},
+		{
+			name: "GetAll evict expired entry",
+			execute: func(c imcache[string, string]) {
+				c.Set("bar", "foo", WithNoExpiration())
+				c.Set("foo", "bar", WithExpirationDate(time.Now().Add(-1*time.Second)))
+				c.GetAll()
+			},
+		},
+	}
+	caches := []struct {
+		name   string
+		create func(EvictionCallback[string, string]) imcache[string, string]
+	}{
+		{
+			name: "Cache",
+			create: func(ec EvictionCallback[string, string]) imcache[string, string] {
+				return New[string, string](WithEvictionCallbackOption[string, string](ec), WithMaxEntriesOption[string, string](2))
+			},
+		},
+		{
+			name: "Sharded",
+			create: func(ec EvictionCallback[string, string]) imcache[string, string] {
+				return NewSharded[string, string](2, DefaultStringHasher64{}, WithEvictionCallbackOption[string, string](ec), WithMaxEntriesOption[string, string](2))
+			},
+		},
+	}
+	for _, cache := range caches {
+		for _, tt := range tests {
+			t.Run(cache.name+" "+tt.name, func(t *testing.T) {
+				ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+				defer cancel()
+				ec := &longRunningEvictionCallback{ctx: ctx}
```

Comment:
Reusing same context can lead to subtle races. Use other cancelation mechanism to end a callback goroutine.